### PR TITLE
Trace ssh sessions

### PR DESCRIPTION
**Pull request by rosstimothy**
Adds a wrapper around `ssh.Session` which injects tracing context
in a similar manner to the [`ssh.Client`](https://github.com/gravitational/teleport/pull/14124) wrapper. 
All usages of `ssh.Session` have now been replaced and have the appropriate
`context.Context` passed along

Part of #12241

### CODE REVIEW

File: api/observability/tracing/ssh/client.go [hunk 1]
```diff
@@ -249,8 +255,172 @@ func (c *Client) NewSession(ctx context.Context) (*ssh.Session, error) {
...
+		oteltrace.WithSpanKind(oteltrace.SpanKindClient),
+		oteltrace.WithAttributes(
+			append(
+				peerAttr(c.Conn.RemoteAddr()),
+				semconv.RPCServiceKey.String("ssh.Client"),
+				semconv.RPCMethodKey.String("OpenChannel"),
+				semconv.RPCSystemKey.String("ssh"),
+			)...,
+		),
+	)
```
Comment (ryanclark) [inline api/observability/tracing/ssh/client.go:375]: This is duplicated quite a few times, with the only things changing are what's in line #365 and #371, and some having extra attributes - whats your thoughts on abstracting it out?

Something like

```go
func StartSSHTrace(ctx context.Context, kind, name string, tracer *tracing.Provider, attributes ...attribute.KeyValue) (*context.Context, trace.error) {
	return tracer.Start(
		ctx,
		fmt.Sprintf("ssh.%s/%s", kind, name),
		oteltrace.WithSpanKind(oteltrace.SpanKindClient),
		oteltrace.WithAttributes(
			append(
				semconv.RPCServiceKey.String("ssh.Client"),
				semconv.RPCMethodKey.String(kind),
				semconv.RPCSystemKey.String("ssh"),
				attributes...,
			)...,
		),
	)
}
```

Comment (ryanclark) [reply api/observability/tracing/ssh/client.go:375]: Yeah I was hesitant to suggest it, as I realise it's still pretty verbose - I was just thinking of keeping similar attributes the same across the board. Happy either way!

File: api/observability/tracing/ssh/client.go [hunk 2]
```diff
@@ -249,8 +255,172 @@ func (c *Client) NewSession(ctx context.Context) (*ssh.Session, error) {
...
+	ssh.Conn
+	// capability the tracingCapability of the ssh server
+	capability tracingCapability
+	// ctx the context which should be used to create spans from
+	ctx context.Context
+	// opts the tracing options to use for creating spans with
+	opts []tracing.Option
+
+	// lock protects the context queue
+	lock sync.Mutex
```
Comment (ryanclark) [inline api/observability/tracing/ssh/client.go:290]: This could be a `sync.RWMutex`

File: api/observability/tracing/ssh/client.go [hunk 3]
```diff
@@ -249,8 +255,172 @@ func (c *Client) NewSession(ctx context.Context) (*ssh.Session, error) {
...
+	defer c.lock.Unlock()
+
+	c.contexts[name] = append(c.contexts[name], ctx)
+}
+
+// nextContext returns the first context.Context for the provided
+// channel name
+func (c *clientWrapper) nextContext(name string) context.Context {
+	c.lock.Lock()
+	defer c.lock.Unlock()
```
Comment (ryanclark) [inline api/observability/tracing/ssh/client.go:342]: & then this becomes

```suggestion
	c.lock.RLock()
	defer c.lock.RUnlock()
```

**Review by ryanclark — APPROVED** [review]

File: api/observability/tracing/ssh/client.go (hunk 3 above)
Comment (fspmarshall) [reply api/observability/tracing/ssh/client.go:342]: This function modifies `c.contexts`.

File: api/observability/tracing/ssh/client.go (hunk 2 above)
Comment (fspmarshall) [reply api/observability/tracing/ssh/client.go:290]: RWMutexes aren't more efficient unless you expect fairly high numbers of concurrent reads.  Near as I can tell that isn't true for this type.

**Review by fspmarshall — APPROVED** [review]
LGTM.  I think if we add any _more_ instrumentation to the ssh types we might want to consider creating interfaces to abstract over them.  Feels fiddly to have to change the types all over to add a wrapper.

**Pull request merged**